libs/s25main/figures/nofMiner.cpp

```diff
@@ -71,20 +72,100 @@ helpers::OptionalEnum<GoodType> nofMiner::ProduceWare()
 
 bool nofMiner::AreWaresAvailable() const
 {
+    // FindPointWithResource triggeres outofresource message
+    if(GetAddonSetting() == 3)
+        return true;
+
     return nofWorkman::AreWaresAvailable() && FindPointWithResource(GetRequiredResType()).isValid();
 }
 
-bool nofMiner::StartWorking()
+unsigned int nofMiner::GetAddonSetting() const 
 {
-    MapPoint resPt = FindPointWithResource(GetRequiredResType());
-    if(!resPt.isValid())
-        return false;
     const GlobalGameSettings& settings = world->GetGGS();
-    bool inexhaustibleRes = settings.isEnabled(AddonId::INEXHAUSTIBLE_MINES)
-                            || (workplace->GetBuildingType() == BuildingType::GraniteMine
-                                && settings.isEnabled(AddonId::INEXHAUSTIBLE_GRANITEMINES));
-    if(!inexhaustibleRes)
-        world->ReduceResource(resPt);
+
+    switch(workplace->GetBuildingType())
+    {
+        case BuildingType::GoldMine: return settings.getSelection(AddonId::MINES_GOLD);
+        case BuildingType::IronMine: return settings.getSelection(AddonId::MINES_IRON);
+        case BuildingType::CoalMine: return settings.getSelection(AddonId::MINES_COAL);
+        case BuildingType::GraniteMine: return settings.getSelection(AddonId::MINES_GRANITE);
+        default: return 0;
+    }
+}
+
+bool nofMiner::StartWorking()
+{
+    workplace->is_emptyCycle = false;
+
+    // needs to have at least one resource spot
+    // 0 = is exhaustible
+    // 1 = settlers IV like
+    // 2 = inexhaustible
+    // 3 = everywhere
+    unsigned int addonSettings = GetAddonSetting();
+
+    switch (addonSettings)
+    {
+        case 1: // settlers IV style
+        {
+            int sumResAmount = 0;
+            MapPoint useResPt;
+
+            std::vector<MapPoint> resPts = FindAllPointsWithResource(GetRequiredResType());
+
+            // iterate over all points
+            for each(MapPoint curPt in resPts)
+            {
+                // calculate the absolute amount of resource beneath
+                uint8_t resAmount = world->GetNode(curPt).resources.getAmount();
+                sumResAmount += resAmount;
+
+                // if there is one with more than 1 quantity, keep it (for reducing)
+                if(resAmount > 1 && !useResPt.isValid())
+                    useResPt = curPt;
+            }
+
+            // no resources left (mine was built on invalid spot)
+            if(sumResAmount == 0)
+                return false;
+
+            // 19 = amount of nodes a mine can reach
+            // 7  = maximum resource amount a node possibly has
+            if(RANDOM.Rand(RANDOM_CONTEXT(), 19 * 7) > sumResAmount)
+            {
+                // failed, use food and start working - but produce nothing
+                workplace->is_emptyCycle = true;
+            } else
+            {
+                // if success, use 1 quantity if any
+                if(!useResPt.isValid())
```

Comment:
This doesn't make sense. On success this must be set, so the `if` can be removed. Maybe even use this check instead of `if(sumResAmount == 0)`?

Reply:
In the Settlers IV like version resources are used up to 1 remaining ore - therefore if all nodes are used up, then no node with > 1 ore is left - so nothing to use and minimum chance for that mine at that spot. Hope that makes sense in a way

Reply:
Ah my bad. Haven't seen that "useResPt" is set to the first node containing **more** than 1 resource.
Maybe improve the comment: I got confused by "use 1 quantity if any" which reads as if: "use 1 res if there is any res anywhere". Suggestion: "Success: Use 1 quantity if any point with more than 1 was found (never fully exhaust the resource)"

The "if success" also made me think it refers to the following condition

Reply:
Ah, you were right, this didn't make sense at alle - if a valid point has been found and the "roll" was successfull, the resource should be used - it was the other way round, so you were right